### PR TITLE
fix overflow bug in prevpow

### PR DESCRIPTION
**Pull request by oscardssmith**
Previously `prevpow(10, 1234567890123456789)` gave `-8446744073709551616` due to integer overflow.

### CODE REVIEW

**Comment by vchuravy** [comment]
Needs a test

File: base/intfuncs.jl [hunk 1]
```diff
@@ -490,14 +490,15 @@ julia> prevpow(4, 16)
...
+function prevpow(a::Real, x::T) where T <: Real
     x < 1 && throw(DomainError(x, "`x` must be ≥ 1."))
     # See comment in nextpos() for a == special case.
     a == 2 && isa(x, Integer) && return _prevpow2(x)
     a <= 1 && throw(DomainError(a, "`a` must be greater than 1."))
     n = floor(Integer,log(a, x))
-    p = a^(n+1)
-    p <= x ? p : a^n
+    p = a^n
+    wp = widemul(p,a)
```
Comment (Keno) [inline base/intfuncs.jl:500]: This doesn't really need the widemul, only the overflow bit (which is free). Does LLVM do the appropriate transformation?

Comment (oscardssmith) [reply base/intfuncs.jl:500]: I could fix this better, but given that I removed an integer pow, this will remain faster. If you have a good way to get the overflow bit, feel free to edit the PR. Otherwise I'll probably merge as is.

Comment (chriselrod) [reply base/intfuncs.jl:500]:
```julia
julia> Base.mul_with_overflow(9,10)
(90, false)

julia> Base.mul_with_overflow(9,typemax(Int))
(9223372036854775799, true)
```

Comment (oscardssmith) [reply base/intfuncs.jl:500]: This can't use `mul_with_overflow` since it needs to work with non-integers.

Comment (KristofferC) [reply base/intfuncs.jl:500]: Just branch on the type of `a`?

File: base/intfuncs.jl [hunk 2]
```diff
@@ -490,14 +490,19 @@ julia> prevpow(4, 16)
...
-function prevpow(a::Real, x::Real)
+function prevpow(a::Real, x::T) where T <: Real
     x < 1 && throw(DomainError(x, "`x` must be ≥ 1."))
     # See comment in nextpos() for a == special case.
     a == 2 && isa(x, Integer) && return _prevpow2(x)
     a <= 1 && throw(DomainError(a, "`a` must be greater than 1."))
     n = floor(Integer,log(a, x))
-    p = a^(n+1)
-    p <= x ? p : a^n
+    p = T(a)^n
```
Comment (simeonschaub) [inline base/intfuncs.jl:499]: `a` might not necessarily convertible to `T`, so I don't think this is correct.

File: base/intfuncs.jl [hunk 3]
```diff
@@ -490,14 +490,19 @@ julia> prevpow(4, 16)
...
+    if a isa Integer && x isa Integer
+        wp, overflow = mul_with_overflow(T(a), x)
```
Comment (simeonschaub) [inline base/intfuncs.jl:501]:
```suggestion
    if p isa Integer && x isa Integer
        wp, overflow = mul_with_overflow(p, oftype(p, x))
```

**Review by simeonschaub — CHANGES_REQUESTED** [review]

File: base/intfuncs.jl (hunk 3 above)
Comment (oscardssmith) [reply base/intfuncs.jl:501]: This is wrong. `wp` needs to be of the same type as `x` to prevent spurious overflow.

Comment (simeonschaub) [reply base/intfuncs.jl:501]: In that case maybe restrict the signature to `prevpow(a::T, x::T) where {T}` and use `prevpow(promote(a, x)...)` for the fallback definition.

Comment (oscardssmith) [reply base/intfuncs.jl:501]: I think that's probably the best answer.

Comment (oscardssmith) [reply base/intfuncs.jl:501]: wait, that doesn't work. The desired behavior is to follow the eltype of the 2nd argument, not the promotion.

Comment (oscardssmith) [reply base/intfuncs.jl:501]: nevermind, it's actually the first argument. `prevpow` is a weird function.

**Comment by simeonschaub** [comment]
I think this is still not fully correct since `x` isn't necessarily the same type as `a`, so `x` could be greater than `typemax(a)` which means that `p*a` overflowing isn't necessarily sufficient here. So perhaps we should still promote `a` and `x` first?

**Comment by oscardssmith** [comment]
That would be a somewhat breaking change since it would change the return type.

**Comment by oscardssmith** [comment]
TLDR is I think that the correct behavior here is either this or to throw an error, which we can't do easily (since there is no `Base.checked_pow`). As such, I think this is what we should merge since it is a strict improvement.

**Comment by oscardssmith** [comment]
@simeonschaub do you think this is mergable as is?

**Comment by KristofferC** [comment]
Might a well do a pkgeval?

**Comment by oscardssmith** [comment]
@nanosoldier runtests(ALL, vs = ":master")

**Comment by oscardssmith** [comment]
`@nanosoldier runtests(ALL, vs = ":master")`

**Comment by KristofferC** [comment]
`@nanosoldier runtests(ALL, vs = ":master")`

**Comment by oscardssmith** [comment]
`@nanosoldier runtests(ALL, vs = ":master")`

**Comment by KristofferC** [comment]
`@nanosoldier runtests(ALL, vs = ":master")`

**Comment by oscardssmith** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/43410#issuecomment-995860127) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_hash/bbef44f_vs_3a3cf02/report.md).

**Comment by KristofferC** [comment]
Looks ok.

**Comment by oscardssmith** [comment]
Merging monday sans objections
